### ISSUE TITLE
Make Client#connect() return a Thenable

### ISSUE THREAD

**Issue by felixfbecker** [brianc/node-postgres#1117]
It doesn't return anything, so it might aswell return a promise

**Comment by vitaly-t** [comment]
All those promise patches aren't doing any good here.

Any recognizable promise library today works like this: if you do not provide a `.catch` section, and the returned promise rejects, the promise library MUST throw an error telling you that the reject hasn't been handled.

This is why returning an optional promise is never a good idea. The result must be either promise-based or a callback-based, with explicit protocol.

@brianc this is why trying to mix callbacks with optional promises isn't a good idea.

**Comment by felixfbecker** [comment]
@vitaly-t there are people out there that use simple native promises. If you want a warning as a user, you can wrap it with `Promise.resolve()` from your favourite promise library.

The decision to add promise support piece-by-piece to pg [has already been made](https://github.com/brianc/node-postgres/blob/master/CHANGELOG.md#v510). The `query()` methods already returns a Thenable, which allows us to directly use it with `async`/`await` already. But for `connect`, we still have to do
```ts
await new Promise((res, rej) => client.connect(err => err ? rej(err) : res()));
```
I know about pg-promise, but 1. it is a wrapper module with an explicit dependency that will always be behind pg (it is one major version behind right now) and 2. the initialization API is completely different from pg and it seems to me that the library also adds a whole lot of additional API that I don't need and is different from pg. For example, why the named parameters with `${}` when you can use actual template strings. Why a transaction API when I have `try`/`catch`/`finally`.

The only thing that is hindering here is that pg needs to support node 0.8, so we could just return a thenable like this:
```js
return {
  then: function then(onFulfilled, onRejected) {
    cb = function(err) {
      if (err) onRejected(err) else onFulFilled()
    }
  }
}
```

7 lines of code that would allow us to use `connect()` with `async`/`await` or any promise library we want.

**Comment by vitaly-t** [comment]
@felixfbecker 

>The decision to add promise support piece-by-piece to pg has already been made

I'm well aware of that :)

>The query() methods already returns a Thenable, which allows us to directly use it with async/await already

That's still an incomplete implementation. A complete one should also work with ES6 generators directly (pg-promise does). This is especially important now that `await/async` are no longer part of ES7.

> it is a wrapper module with an explicit dependency that will always be behind pg

It has never been behind `pg`, always using its best (in terms of reliability) version. The reason it is currently using `5.1` is because starting from `5.2` the new connection pool was introduced which is known to fail in repairing broken database connections - one of the most important aspects in real-world usage of the framework. Once this issue has been fixed, I will upgrade pg-promise right away.

> the initialization API is completely different from pg

Is it really? It is using the exact same connection string + config object to configure the connection, passing it directly into pg without any modification.

>also adds a whole lot of additional API that I don't need and is different from pg

It adds higher-level API, to make it easier to use, which you don't have to use, if you do not need to. How it is different, again, it isn't really, it has all the same methods in it, plus some extras.

>why the named parameters with ${} when you can use actual template strings

Syntax `${}` is supported as a legacy, since it was initially written before ES6 existed. And it is optional, with alternatives available. And you can't use template strings to escape query values.

>Why a transaction/"task" API when I have try/catch/finally.

What does it even mean? Transactions got nothing to do with `try/catch/finally`.

>The only thing that is hindering here is that pg needs to support node 0.8

It doesn't, actually.  pg supports only 0.10 upwards, as per [travis ci](https://github.com/brianc/node-postgres/blob/master/.travis.yml).

P.S. I won't be following up here on anything related to pg-promise, as it is not the right place for it. If you have questions - you can them on pg-promise page or stackoverflow ;)

**Comment by felixfbecker** [comment]
@vitaly-t 
> That's still an incomplete implementation. A complete one should also work with ES6 generators directly (pg-promise does).

What do you mean by that? It is not a complete _promise_ implementation. But it is a _Thenable_ per spec as it implements a `.then()` method that accepts two callbacks for success and error. This makes it possible to use it with `Promise.resolve()`, `async`/`await`, `co` and `Bluebird.coroutine()` if that's what you are referring too with generators.

> This is especially important now that await/async are no longer part of ES7.

"ES7" doesn't even exist, ECMAScript will see yearly releases and async/await [is a Stage 3 proposal](https://tc39.github.io/ecmascript-asyncawait/) already implemented in Chakra, currently being implemented in V8 and available for all Babel and TypeScript users.

> And you can't use template strings to escape query values.

I can very well with tagged template strings, see [sql-template-strings](https://www.npmjs.com/package/sql-template-strings).

> Transactions got nothing to do with try/catch/finally

```js
await client.query('BEGIN')
try {
  await client.query('whatever')
  await client.query('COMMIT')
} catch (err) {
  await client.query('ROLLBACK')
  throw err;
}
```

**Comment by vitaly-t** [comment]
> and `Bluebird.coroutine()` if that's what you are referring too with generators.

What I was referring to, is not having to rely on `Bluebird.coroutine()`, specific to that one library, but to be able to use with any promise library.

>"ES7" doesn't even exist

The initial spec was finalized about 1-2 month ago, with `async/await` excluded from it. I won't speculate on further revisions down the road, they may as well become ES8/9.

I am familiar with both babel and TypeScript, but most users of the pg library don't use those.

>I can very well with tagged template strings

Ha-ha, you are writing your own query string formatting. Well done, but I think we should park our discussion in that case right here, or it may take forever.

**Comment by felixfbecker** [comment]
> The initial spec was finalized about 1-2 month ago, with async/await excluded from it.

what I meant was that there are now simply called ES2016 / 17 / 18 etc. In the end what matters is not the spec but if V8 implements it.

> Ha-ha, you are writing your own query string formatting. Well done, but I think we should park our discussion in that case right here, or it may take forever.

To be fair, I actually got the idea from the pg Wiki, were it was a simple snippet using `array.reduce()`. I just turned it cross-dialect and published it as a module.

Your points about transactions are all valid but my initial statement was simply that I don't need all this, I am fine with pg, I just feel that the API would really benefit from this. This isn't just about high-level code, sequelize for example uses promises in the whole codebase and uses pg aswell, so we could simplify code there.

> Honestly, I don't care so much about this library getting some half-baked promise support. This is the base driver, and as such its focus should be on reliability and integration simplicity, not on simplifying direct usage - the frameworks built on it do that.
> 
> I am frustrated I can't upgrade pg-promise to the 6.x branch because of the issue introduced with broken connections. So when I see the library getting needless features instead of fixing the priority issue, it is kind of upsetting. Looks like I'm gonna be using 5.1 for quite some time, as it is the last solid version of this library.

I can understand this, I didn't want to demand the maintainers to implement this, rather hear how open they would be to a PR 🙂

**Comment by brianc** [comment]
1) Client.query.then in node-postgres works AFAIK with generators. I use it with `co` all the time. If it doesn't work please file an issue.

2) @vitaly-t from now on if you mention pg-promise in an issue report here I will delete your comment. If you continue I will block you.  I've felt for a long time like you're more interested in building your personal brand than providing meaningful code contributions back to this library.  You're not even in the top 10 contributors, and the bulk of your issue comments you eventually find some way to slip in some marketing for pg-promise. I'm glad you wrote that library, and I wish you all the success with it, but stop with the constant self-promotion. If you think 5.1 is "the last solid version of this library" than open up a pull request to fix whatever perceived changes are causing issues. Sniping and complaining without solutions only causes me to get more and more burned out.

3) I added promises because I wanted to...[that's how open source works](https://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/).

5) @felixfbecker - I'd definitely be down with adding a `.thenable` thing to other areas so the api will support promises and callbacks in tandem.  It's a bit of a tricky dance to get right & I've probably made a few mistakes so if you catch any I'm happy to help fix 'em, review PRs, all that stuff!  The reason I did a kinda 1/2 promise return instead of actually returning an instance of `new Promise` is because I want this library to still support node `v 0.10.x` until the next LTS version of node is released, and `v 0.10.x` doesn't have a native promise implementation.  So I kinda rolled a `.thenable` in place here.  In pg-pool I expect a promise to exist so I'm open to changing how all this works.  Maybe open a PR with the change to adding a promise/thenable/whatever to client.connect & we can continue to discuss there?
